[Feature] Add tile-based method of supporting large VPT in moe_fused_gate kernel

ltaodream opened the pull request:
## Motivation

The kimi-vl supported at #5383 and kimi-k2 model cannot use the fuse moe gate operator. This operator currently does not support the case where the group expert is greater than 32. The group expert of kimi-vl is 64, the group expert of kimi-k2 is 384, and this operator cannot be used at present.

## Modifications

@ttaohe @Misaka9468 The three of us completed the PR "Add tile-based method of supporting large VPT in moe_fused_gate kernel" together.

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

ltaodream committed:
add support the number of group expert from the original 32 to 128 implemented with tilling

ltaodream committed:
modify the MAX_VPT limit in topk.py

ltaodream committed:
Fix the size of row_chunk and bias_chunk to 32 to prevent register overflow

Alcanderian commented on sgl-kernel/csrc/moe/moe_fused_gate.cu:
```diff
@@ -81,57 +82,107 @@ __device__ void moe_fused_gate_impl(
+          } else if (cmp_gt(val, global_max_val_second)) {
+            global_max_val_second = val;
+            global_max_second_idx = global_idx;
+          }
```
I cannot see the difference between these two block

yuan-luo replied:
Agree, there are a lot of duplicated code which can be refactored.

Alcanderian commented on sgl-kernel/csrc/moe/moe_fused_gate.cu:
```diff
@@ -81,57 +82,107 @@ __device__ void moe_fused_gate_impl(
+          } else if (cmp_gt(val, global_max_val_second)) {
+            global_max_val_second = val;
+            global_max_second_idx = global_idx;
+          }
```
I cannot see the difference between these two block

ltaodream committed:
add test and cycle fusion Optimization

ltaodream committed:
add bench_moe_fused_gate for kimi-vl

BBuf committed:
Merge branch 'main' into main

ltaodream committed:
Merge branch 'sgl-project:main' into main

ltaodream committed:
[Kimi K2] add support the number of group expert from the original 32 to 384 implemented with tilling

BBuf committed:
Merge branch 'main' into main

ltaodream committed:
fix prefetch_global、fix alignment error with cutlass alignedarray

ltaodream committed:
Merge branch 'main' of ssh://github.com/ltaodream/sglang

BBuf commented on sgl-kernel/csrc/moe/moe_fused_gate.cu:
```diff
@@ -492,56 +501,83 @@ std::vector<at::Tensor> moe_fused_gate(
   switch (num_experts) {
+    case 384:
+      if (num_expert_group == 1)
+        // Kimi K2 配置: VPT = 384/1 = 384, ROWS_PER_WARP = 32/1 = 32
```
// Kimi K2 config:

ltaodream committed:
fix prefetch

ltaodream committed:
fix kimi k2 config

ltaodream committed:
feat: Optimize MoE routing with efficient shared memory CUDA kernel

ltaodream committed:
test: Enhance MoE routing tests with comprehensive group selection verification

ltaodream committed:
add test_moe_fused_gate param

ltaodream committed:
Merge branch 'sgl-project:main' into main

ltaodream committed:
fix: Adjust MOE routing test tolerances for different expert configurations

ltaodream committed:
Merge branch 'main' of ssh://github.com/ltaodream/sglang

BBuf committed:
Merge branch 'main' into main

ltaodream committed:
fix: Optimize MOE Fused Gate kernel to use non-tiled path for VPT≤32 and tiled path for VPT>32

ltaodream committed:
Merge branch 'main' of ssh://github.com/ltaodream/sglang

BBuf commented on sgl-kernel/tests/test_moe_fused_gate.py:
```diff
@@ -38,6 +40,7 @@ def test_moe_fused_gate_combined(seq_length, params, num_fused_shared_experts):
         routed_scaling_factor=2.5,
     )
+    
     ref_output, ref_indices = biased_grouped_topk(
```
We must change this function to use `biased_grouped_topk_impl` from `sglang.srt.layers.moe.topk` to avoid self-comparison.

```suggestion
from sglang.srt.layers.moe.topk import biased_grouped_topk_impl
    ref_output, ref_indices = biased_grouped_topk_impl(
```

ltaodream replied:
This has been modified in the latest submission.

ltaodream committed:
fix: correct function call from biased_grouped_topk to biased_grouped_topk_impl

ltaodream committed:
feat: use native topk for single expert group with group size > 128, tile-based kernel for other cases

ltaodream commented:
> Add benchmark results with `sglang/sgl-kernel/benchmark/bench_moe_fused_gate.py`.

Fixed the bug in bench and test: changed biased_grouped_topk to biased_grouped_topk_impl

The performance comparison between the original moe_fused_gate.cu and PyTorch's native implementation is as follows. It can be observed that the scenario (sq, 256, 8, 4, 8) performs well, but it does not support single-group expert counts of 64 and 384.

```
seq_length_range = [5000, 10000, 15000, 20000, 25000, 30000, 35000, 40000]
configs = [(sq, 256, 8, 4, 8) for sq in seq_length_range] 
moe-fused-gate-performance:
   seq_length     Original  SGL Kernel
0      5000.0   235.551998   31.808000
1     10000.0   362.511992   38.015999
2     15000.0   477.151990   47.040001
3     20000.0   626.496017   53.119998
4     25000.0   776.416004   60.959999
5     30000.0   892.960012   65.728001
6     35000.0  1042.352021   71.712002
7     40000.0  1158.176005   79.967998
```

```
seq_length_range = [5000, 10000, 15000, 20000, 25000, 30000, 35000, 40000]
configs = ([(sq, 64, 1, 1, 6) for sq in seq_length_range])  # kimi vl config
Traceback (most recent call last):
  File "/home/yineng/bbuf/sglang/sgl-kernel/benchmark/bench_moe_fused_gate.py", line 76, in <module>
    benchmark.run(print_data=True)
  File "/usr/local/lib/python3.10/dist-packages/triton/testing.py", line 392, in run
    result_dfs.append(self._run(bench, save_path, show_plots, print_data, **kwargs))
  File "/usr/local/lib/python3.10/dist-packages/triton/testing.py", line 335, in _run
    ret = self.fn(**x_args, **{bench.line_arg: y}, **bench.args, **kwrags)
  File "/home/yineng/bbuf/sglang/sgl-kernel/benchmark/bench_moe_fused_gate.py", line 65, in benchmark
    ms, min_ms, max_ms = triton.testing.do_bench(
  File "/usr/local/lib/python3.10/dist-packages/triton/testing.py", line 145, in do_bench
    fn()
  File "/home/yineng/bbuf/sglang/sgl-kernel/benchmark/bench_moe_fused_gate.py", line 66, in <lambda>
    lambda: biased_grouped_topk_org_fuse_kernel(
  File "/home/yineng/bbuf/sglang/sgl-kernel/benchmark/bench_moe_fused_gate.py", line 27, in biased_grouped_topk_org_fuse_kernel
    return moe_fused_gate(scores, bias, num_expert_group, topk_group, topk)
  File "/usr/local/lib/python3.10/dist-packages/sgl_kernel/moe.py", line 58, in moe_fused_gate
    return torch.ops.sgl_kernel.moe_fused_gate.default(
  File "/usr/local/lib/python3.10/dist-packages/torch/_ops.py", line 756, in __call__
    return self._op(*args, **kwargs)
RuntimeError: Per group experts: num_experts / num_expert_group = (64) exceeds the maximum supported (32)
```

```
seq_length_range = [5000, 10000, 15000, 20000, 25000, 30000, 35000, 40000]
configs = ([(sq, 384, 1, 1, 8) for sq in seq_length_range])  # kimi k2 config

➜  benchmark git:(main) ✗ python bench_moe_fused_gate.py
Traceback (most recent call last):
  File "/home/yineng/bbuf/sglang/sgl-kernel/benchmark/bench_moe_fused_gate.py", line 76, in <module>
    benchmark.run(print_data=True)
  File "/usr/local/lib/python3.10/dist-packages/triton/testing.py", line 392, in run
    result_dfs.append(self._run(bench, save_path, show_plots, print_data, **kwargs))
  File "/usr/local/lib/python3.10/dist-packages/triton/testing.py", line 335, in _run
    ret = self.fn(**x_args, **{bench.line_arg: y}, **bench.args, **kwrags)
  File "/home/yineng/bbuf/sglang/sgl-kernel/benchmark/bench_moe_fused_gate.py", line 65, in benchmark
    ms, min_ms, max_ms = triton.testing.do_bench(
  File "/usr/local/lib/python3.10/dist-packages/triton/testing.py", line 145, in do_bench
    fn()
  File "/home/yineng/bbuf/sglang/sgl-kernel/benchmark/bench_moe_fused_gate.py", line 66, in <lambda>
    lambda: biased_grouped_topk_org_fuse_kernel(
  File "/home/yineng/bbuf/sglang/sgl-kernel/benchmark/bench_moe_fused_gate.py", line 27, in biased_grouped_topk_org_fuse_kernel
    return moe_fused_gate(scores, bias, num_expert_group, topk_group, topk)
  File "/usr/local/lib/python3.10/dist-packages/sgl_kernel/moe.py", line 58, in moe_fused_gate
    return torch.ops.sgl_kernel.moe_fused_gate.default(
  File "/usr/local/lib/python3.10/dist-packages/torch/_ops.py", line 756, in __call__
    return self._op(*args, **kwargs)
RuntimeError: num_experts must be a power of 2, but got 384
```

The performance comparison between the modified moe_fused_gate.cu and the native PyTorch implementation is as follows:

```
seq_length_range = [5000, 10000, 15000, 20000, 25000, 30000, 35000, 40000]
configs = ([(sq, 64, 1, 1, 6) for sq in seq_length_range])  # kimi vl config
moe-fused-gate-performance:
   seq_length  num_experts  num_expert_group  topk_group  topk    Original  SGL Kernel
0      5000.0         64.0               1.0         1.0   6.0  217.600003   72.352000
1     10000.0         64.0               1.0         1.0   6.0  212.896004   75.680003
2     15000.0         64.0               1.0         1.0   6.0  218.431994   76.223999
3     20000.0         64.0               1.0         1.0   6.0  211.968005   75.231999
4     25000.0         64.0               1.0         1.0   6.0  220.927998   78.975998
5     30000.0         64.0               1.0         1.0   6.0  255.008012   86.336002
6     35000.0         64.0               1.0         1.0   6.0  288.143992   87.871999
7     40000.0         64.0               1.0         1.0   6.0  313.055992   85.536003
```

```
seq_length_range = [5000, 10000, 15000, 20000, 25000, 30000, 35000, 40000]
configs = [(sq, 256, 8, 4, 8) for sq in seq_length_range] 

moe-fused-gate-performance:
   seq_length  num_experts  num_expert_group  topk_group  topk     Original  SGL Kernel
0      5000.0        256.0               8.0         4.0   8.0   213.632002   31.968001
1     10000.0        256.0               8.0         4.0   8.0   362.399995   38.559999
2     15000.0        256.0               8.0         4.0   8.0   477.295995   47.456000
3     20000.0        256.0               8.0         4.0   8.0   624.655992   54.240000
4     25000.0        256.0               8.0         4.0   8.0   776.287973   62.080000
5     30000.0        256.0               8.0         4.0   8.0   890.415996   67.456000
6     35000.0        256.0               8.0         4.0   8.0  1042.559981   72.927997
7     40000.0        256.0               8.0         4.0   8.0  1158.079982   81.039999
```

```
seq_length_range = [5000, 10000, 15000, 20000, 25000, 30000, 35000, 40000]
configs = ([(sq, 384, 1, 1, 8) for sq in seq_length_range]) 
moe-fused-gate-performance:
   seq_length  num_experts  num_expert_group  topk_group  topk     Original  SGL Kernel
0      5000.0        384.0               1.0         1.0   8.0   224.992007  149.535999
1     10000.0        384.0               1.0         1.0   8.0   386.848003  241.600007
2     15000.0        384.0               1.0         1.0   8.0   556.272000  329.712003
3     20000.0        384.0               1.0         1.0   8.0   715.807974  420.064002
4     25000.0        384.0               1.0         1.0   8.0   884.000003  509.184003
5     30000.0        384.0               1.0         1.0   8.0  1044.543982  598.095983
6     35000.0        384.0               1.0         1.0   8.0  1216.256022  687.904000
7     40000.0        384.0               1.0         1.0   8.0  1374.800026  775.887996
```

```
seq_length_range = [5000, 10000, 15000, 20000, 25000, 30000, 35000, 40000]
configs = [(sq, 384, 8, 4, 8) for sq in seq_length_range] 
moe-fused-gate-performance:
   seq_length  num_experts  num_expert_group  topk_group  topk     Original  SGL Kernel
0      5000.0        384.0               8.0         4.0   8.0   288.448006  108.032003
1     10000.0        384.0               8.0         4.0   8.0   510.399997  114.848003
2     15000.0        384.0               8.0         4.0   8.0   706.639975  180.383995
3     20000.0        384.0               8.0         4.0   8.0   927.616000  176.479995
4     25000.0        384.0               8.0         4.0   8.0  1163.200021  217.375994
5     30000.0        384.0               8.0         4.0   8.0  1343.232036  255.695999
6     35000.0        384.0               8.0         4.0   8.0  1576.000035  274.239987
7     40000.0        384.0               8.0         4.0   8.0  1768.512011  325.632006
```

ltaodream committed:
fix

BBuf committed:
Merge branch 'main' into main

ispobock reviewed (changes requested):
Please fix the lint issues: 
https://github.com/sgl-project/sglang/actions/runs/16552362162/job/46855986121?pr=6946
https://github.com/sgl-project/sglang/actions/runs/16552362176/job/46855986713?pr=6946

ltaodream committed:
fix lint issues

ltaodream committed:
Merge branch 'main' of ssh://github.com/ltaodream/sglang

ltaodream commented:
> It seems there's a performance drop with this setting?
> 
> ```
> seq_length_range = [5000, 10000, 15000, 20000, 25000, 30000, 35000, 40000]
> configs = [(sq, 256, 8, 4, 8) for sq in seq_length_range]
> ```

You should look at the SGL Kernel column.

ltaodream committed:
Merge branch 'main' into main

ltaodream committed:
fix lint

ltaodream committed:
Merge branch 'main' of ssh://github.com/ltaodream/sglang

ltaodream committed:
Merge branch 'main' into main

BBuf committed:
Merge branch 'main' into main

yuan-luo commented:
Can you paste the precision compare test result between this kernel and baseline kernel?
Per my test, there are mismatch between this moe_fused_gate kernel(before you fix) and the native version.

ltaodream commented on sgl-kernel/csrc/moe/moe_fused_gate.cu:
```diff
@@ -78,176 +86,475 @@ __device__ void moe_fused_gate_impl(
-    row_chunk[ii] = vec_thread_read_ptr[0][ii];
-    bias_chunk[ii] = vec_bias_thread_read_ptr[0][ii];
-  }
+    for (int ii = 0; ii < params.VPT; ++ii) {
```
This should be caused by using VSCode.

yuan-luo replied:
pre-commit run --all-files

ltaodream commented on sgl-kernel/benchmark/bench_moe_fused_gate.py:
```diff
@@ -29,12 +29,13 @@ def biased_grouped_topk_org_fuse_kernel(
-configs = [(sq,) for sq in seq_length_range]
-
+configs = [(sq, 256, 8, 4, 8) for sq in seq_length_range]  # original config
+# configs = ([(sq, 64, 1, 1, 6) for sq in seq_length_range])  # kimi vl config
```
Here I am uncommenting to benchmark one of the three configurations separately.

yuan-luo commented on sgl-kernel/csrc/moe/moe_fused_gate.cu:
```diff
@@ -78,176 +86,475 @@ __device__ void moe_fused_gate_impl(
-  AccessType<T> const* vec_bias_thread_read_ptr = reinterpret_cast<AccessType<T> const*>(bias_thread_read_ptr);
-
+  // Select the execution path based on the VPT value
+  if (params.VPT <= 32) {
```
There are quite a lot of duplicated code between params.VPT <= 32 and else branches, I think they can be consolidated into one sub kernel, or split them into 2 separated kernels.

ltaodream replied:
I have refactored the code by abstracting the logic for "params.VPT <= 32" and "else" branches into two dedicated device functions (moe_fused_gate_small_vpt and moe_fused_gate_large_vpt), and corresponding kernel wrappers. The dynamic kernel launch logic for all data types and VPT cases is now unified through a single macro, significantly reducing duplication and improving maintainability.

ltaodream replied:
Thanks for your suggestion! I have refactored the code to consolidate the logic for both small and large VPT cases into a unified kernel implementation, minimizing code duplication while keeping the structure clear and maintainable.

ltaodream committed:
Refactor dynamic kernel launch code to remove duplication between small and large VPT branches.

ltaodream committed:
Merge branch 'main' of ssh://github.com/ltaodream/sglang

ltaodream committed:
Refactor dynamic kernel launch logic for moe_fused_gate, unify macro

ltaodream committed:
fix

ltaodream committed:
Recovery

ltaodream committed:
Consolidate duplicated logic for small and large VPT cases into a unified kernel implementation.

ltaodream committed:
fix tile

ltaodream committed:
pre-commit

ltaodream committed:
fix

ltaodream committed:
fix index out of bounds

ltaodream committed:
fix num_expert_group=1

ltaodream committed:
fix Expert repeated selection

ltaodream committed:
pre-commit run

ltaodream committed:
fix clear bias

Commit:
upd moe_fuse_gate_tiling_more_experts v1

Commit:
upd

Commit:
fix grouptopk only set 1-shared_experts problem

Commit:
upd

Commit:
add tiled dynamic benchmark

ttaohe committed:
upd pytest style

ttaohe committed:
add bench_moe_fused_gate_tiled script

ttaohe committed:
add specific examples like kimi-vl and kimi-k2 moe config

ttaohe committed:
add kimi-vl 64

ttaohe committed:
bench tile add more test case

ttaohe committed:
change specific style

ttaohe committed:
commit last bench test

ltaodream committed:
Merge pull request #1 from ttaohe/main

[Fix] Modify tile-based method of supporting large VPT in moe_fused_gate kernel

ltaodream commented:
@Misaka9468@Misaka9468 The three of us completed the PR "Add tile-based method of supporting large VPT in moe_fused_gate kernel" together.


MoE Fused Gate: add tiled path and static specializations for large VPT (64/384), unify switch-case dispatch, and provide multi-dtype benchmarking.

- Added a tiled implementation for large VPT and two static specializations for THREADS_PER_ROW=1: (num_experts=64, group=1) and (num_experts=384, group=1). These are exposed via consistent switch-case dispatch using LAUNCH_MOE_GATE_TILED_CONFIG.
- Kept existing template fast paths for small VPT (e.g., 128/256), and route all other large-VPT cases to the generic tiled kernel.
- Refactored tiled declarations/macros into moe_fused_gate_tiled.h to keep moe_fused_gate.cu clean.
- Added a benchmark (bf16/fp16/fp32) comparing Original (eager, compile-static, compile-dynamic) vs SGL Kernel on representative large-VPT configs.

**Results of kimi-vl and kimi-k2**
<img width="960" height="172" alt="image" src="https://github.com/user-attachments/assets/1e29d5d5-7889-43e5-be15-a1a4e8b41816" />

**Benchmmmu result of kimi-vl**
<img width="618" height="549" alt="image" src="https://github.com/user-attachments/assets/fc0f278b-e293-4aae-992b-d2da487ff814" />

**Bench_hf result of kimi-vl**
```
python benchmark/mmmu/bench_hf.py --model-path moonshotai/Kimi-VL-A3B-Instruct

answers saved to: moonshotai/Kimi-VL-A3B-Instruct_val_hf.json
Evaluating...
{'Accounting': {'acc': 0.4, 'num': 30},
 'Agriculture': {'acc': 0.562, 'num': 16},
 'Architecture_and_Engineering': {'acc': 0.367, 'num': 30},
 'Art': {'acc': 0.5, 'num': 30},
 'Art_Theory': {'acc': 0.3, 'num': 30},
 'Basic_Medical_Science': {'acc': 0.667, 'num': 30},
 'Biology': {'acc': 0.333, 'num': 30},
 'Chemistry': {'acc': 0.2, 'num': 30},
 'Clinical_Medicine': {'acc': 0.4, 'num': 30},
 'Computer_Science': {'acc': 0.6, 'num': 30},
 'Design': {'acc': 0.533, 'num': 30},
 'Diagnostics_and_Laboratory_Medicine': {'acc': 0.367, 'num': 30},
 'Economics': {'acc': 0.533, 'num': 30},
 'Electronics': {'acc': 0.333, 'num': 30},
 'Energy_and_Power': {'acc': 0.4, 'num': 30},
 'Finance': {'acc': 0.4, 'num': 30},
 'Geography': {'acc': 0.533, 'num': 30},
 'History': {'acc': 0.5, 'num': 30},
 'Literature': {'acc': 0.828, 'num': 29},
 'Manage': {'acc': 0.333, 'num': 30},
 'Marketing': {'acc': 0.6, 'num': 30},
 'Materials': {'acc': 0.367, 'num': 30},
 'Math': {'acc': 0.4, 'num': 30},
 'Mechanical_Engineering': {'acc': 0.467, 'num': 30},
 'Music': {'acc': 0.267, 'num': 30},
 'Overall': {'acc': 0.44, 'num': 885},
 'Overall-Art and Design': {'acc': 0.4, 'num': 120},
 'Overall-Business': {'acc': 0.453, 'num': 150},
 'Overall-Health and Medicine': {'acc': 0.453, 'num': 150},
 'Overall-Humanities and Social Science': {'acc': 0.563, 'num': 119},
 'Overall-Science': {'acc': 0.353, 'num': 150},
 'Overall-Tech and Engineering': {'acc': 0.434, 'num': 196},
 'Pharmacy': {'acc': 0.467, 'num': 30},
 'Physics': {'acc': 0.3, 'num': 30},
 'Psychology': {'acc': 0.367, 'num': 30},
 'Public_Health': {'acc': 0.367, 'num': 30},
 'Sociology': {'acc': 0.567, 'num': 30}}
eval out saved to moonshotai/Kimi-VL-A3B-Instruct_val_hf.json
Overall accuracy: 0.44
```


Summary MMMU Benchmark on Kimi-VL-A3B-Instruct Model.
Results as follows:
[bench_sglang.py]: 0.512
[bench_hf.py]: 0.44

ltaodream committed:
pre-commit run --all-files

ltaodream committed:
Merge branch 'main' into main

ltaodream commented:
> Cool, maybe need rebase master.

Hi, I've rebased my branch onto the latest main - the PR is now ready for review. Thanks!

ltaodream commented:
The original PR link has been closed. New PR link: [Feature] Add tile-based method of supporting large VPT in moe_fused_gate kernel by ltaodream · Pull Request #9579 · sgl-project/sglang

ltaodream commented:
> > @FlamingoPg Hi, I’ve rebased my branch onto the latest main. Since the old branch had a long chain of merge commits from the original main, the history is now completely rewritten and GitHub can’t update PR #6946 cleanly. Would you prefer that I **close #6946 and open a fresh PR** with the single squash commit, or do you have another way you’d like to handle it?
> 
> Maybe you can reopen a new PR

New PR link: [Feature] Add tile-based method of supporting large VPT in moe_fused_gate kernel by ltaodream · Pull Request https://github.com/sgl-project/sglang/pull/9579
